src/DOM.js

```diff
@@ -1805,62 +1805,67 @@ class HTMLScriptElement extends HTMLLoadableElement {
 
   isRunnable() {
     const {type} = this;
-    return !type || /^(?:(?:text|application)\/javascript|application\/ecmascript)$/.test(type);
+    return !type || type === 'module' || /^(?:(?:text|application)\/javascript|application\/ecmascript)$/.test(type);
+  }
+
+
+  isModule() {
+    return this.getAttribute('type') === 'module';
   }
 
   loadRunNow() {
     this.readyState = 'loading';
 
-    const url = _mapUrl(this.src, this.ownerDocument.defaultView);
-
     return this.ownerDocument.resources.addResource((onprogress, cb) => {
-      this.ownerDocument.defaultView.fetch(url)
-        .then(res => {
-          if (res.status >= 200 && res.status < 300) {
-            return res.text();
-          } else {
-            return Promise.reject(new Error('script src got invalid status code: ' + res.status + ' : ' + url));
-          }
-        })
-        .then(s => {
-          vm.runInThisContext(s, {
-            filename: url,
+      const _getSrc = () => {
+        const innerHTML = this.childNodes.length > 0 ? this.childNodes[0].value : '';
+        if (innerHTML) {
+          const url = this.ownerDocument.defaultView.location.href;
+          return Promise.resolve({
+            s: innerHTML,
+            url,
           });
-        })
-        .then(() => {
-          this.readyState = 'complete';
+        } else {
+          const url = _mapUrl(this.src, this.ownerDocument.defaultView);
+          return _fetch(url).then(s => ({
+            s,
+            url,
+          }));
+        }
+      };
+      const _fetch = async url => {
+        const res = await this.ownerDocument.defaultView.fetch(url);
+        if (res.status >= 200 && res.status < 300) {
+          return await res.text();
+        } else {
+          throw new Error('script src got invalid status code: ' + res.status + ' : ' + url);
+        }
+      };
 
-          this.dispatchEvent(new Event('load', {target: this}));
+      _getSrc()
+        .then(async ({s, url}) => {
+          const opts = {
+            lineOffset : this.location && this.location.line !== null ? this.location.line - 1 : 0,
+            columnOffset: this.location && this.location.col !== null ? this.location.col - 1 : 0,
+          };
 
-          cb();
+          if (this.getAttribute('type') === 'module') {
+            opts.url = url;
+            const script = new vm.SourceTextModule(s, opts);
+            await script.link(async (url, {url: baseUrl}) => {
+              url = _mapUrl(_normalizeUrl(url, baseUrl), this.ownerDocument.defaultView);
+              const s = await _fetch(url);
+              return new vm.SourceTextModule(s, {
+                url: baseUrl,
```

Comment:
You need to use the parent module's `context` here.

```
await script.link(async (url, {url: baseUrl, context}) => {
  ...
    return new vm.SourceTextModule(s, {
      context,
```

See https://nodejs.org/api/vm.html#vm_class_vm_sourcetextmodule

Reply:
Would that be different from leaving it out and having the context be the global `window` (as intended)?